saleor/graphql/checkout/mutations/utils.py

```diff
@@ -396,23 +396,23 @@ def group_lines_input_on_add(
                 if not line_db_id:
                     line_data = checkout_lines_data_map[variant_db_id]
                     line_data.variant_id = variant_db_id
-                    line_data.metadata_list = metadata_list
+                    line_data.metadata_list = metadata_list_from_input
                 else:
                     line_data = checkout_lines_data_map[line_db_id]
                     line_data.line_id = line_db_id
                     line_data.variant_id = find_variant_id_when_line_parameter_used(
                         line_db_id, existing_lines_info
                     )
 
-                    if line_data.metadata_list and metadata_list:
-                        line_data.metadata_list += metadata_list
-                    else:
-                        line_data.metadata_list = metadata_list
+                    if metadata_list_from_input:
+                        line_data.metadata_list = line_data.metadata_list or []
```

Comment:
Maybe instead of doing this, the default value of `CheckoutLineData` should be changed to `list`? Currently it's `None`, and I suppose it's the root of the issue

Reply:
I'd prefer to avoid another refactoring - it's working. We have introduced MetadataCollection dataclass that should be used to persist and operate on metadata - IMO if we refactor this, it should be to this dataclass

And it will also encapsulate such behavior, eg. `m = MetadataCollection(...); m.appendItems(...)`

Reply:
done